### PR TITLE
Push docker image to registry too

### PR DESCRIPTION
**Pull request by chrissng**
- Also install hadoop binaries as some of our projects uses the `hdfs dfs` command

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -68,27 +66,39 @@ RUN apt-get update \
...
-ARG SPARK_VARIANT
+ARG HADOOP_VERSION
 ARG SPARK_PY4J
 
+ENV HADOOP_HOME /opt/hadoop
+ENV PATH ${PATH}:${HADOOP_HOME}/bin:${HADOOP_HOME}/sbin
+ENV HADOOP_CONF_DIR ${HADOOP_HOME}/etc/hadoop
+ENV HDFS_CONF_DIR ${HADOOP_CONF_DIR}
+ENV YARN_CONF_DIR ${HADOOP_CONF_DIR}
+ENV HIVE_CONF_DIR ${HADOOP_CONF_DIR}
```
Comment (xkjyeah) [inline Dockerfile]: No point specifying these directories, since the user will almost certainly override them

File: Dockerfile [hunk 2]
```diff
@@ -68,27 +66,39 @@ RUN apt-get update \
...
 # Download Spark
 ARG SPARK_EXTRACT_LOC=/sparkbin
-RUN set -eux && \
+RUN set -eux && \\
+    curl http://www-us.apache.org/dist/hadoop/common/hadoop-${HADOOP_VERSION}/hadoop-${HADOOP_VERSION}.tar.gz | \
+    tar -xz -C /opt/ && \
+    mv /opt/hadoop-${HADOOP_VERSION} /opt/hadoop && \
     mkdir -p ${SPARK_EXTRACT_LOC} && \
-    curl https://www.mirrorservice.org/sites/ftp.apache.org/spark/spark-${SPARK_VERSION}/spark-${SPARK_VERSION}-bin-${SPARK_VARIANT}.tgz |\
+    curl https://www.mirrorservice.org/sites/ftp.apache.org/spark/spark-${SPARK_VERSION}/spark-${SPARK_VERSION}-bin-hadoop${HADOOP_VERSION:0:3}.tgz | \
```
Comment (xkjyeah) [inline Dockerfile:87]: Can use the Hadoop-provided variant to keep bundle small. But need to update spark-submit.sh subsequently

File: Dockerfile [hunk 3]
```diff
@@ -68,27 +66,39 @@ RUN apt-get update \
...
     rm -rf ${SPARK_EXTRACT_LOC} && \
     echo SPARK_HOME is ${SPARK_HOME} && \
     ls -al --g ${SPARK_HOME}
 
 # Less verbose logging
 COPY log4j.properties.production ${SPARK_HOME}/conf/log4j.properties
 
-
+## To build your own image:
+ONBUILD COPY hadoop/conf/ ${HADOOP_CONF_DIR}/
```
Comment (xkjyeah) [inline Dockerfile]: I don't know... We could be targeting three or four different systems, I don't see the point of copying in the config files

File: Dockerfile (hunk 2 above)
Comment (xkjyeah) [reply Dockerfile:87]: *-env.sh

**Review by xkjyeah — CHANGES_REQUESTED** [review]

**Comment by xkjyeah** [comment]
Crap I remembered: the default spark distribution does not come with Hive.

That's why I built my own in spark-yarn-cluster.

File: Dockerfile (hunk 2 above)
Comment (chrissng) [reply Dockerfile]: I'm not sure what you are referring to about hadoop provided variant.

File: Dockerfile (hunk 3 above)
Comment (chrissng) [reply Dockerfile]: True that. Will remove this and update the Readme.

**Comment by xkjyeah** [comment]
https://spark.apache.org/docs/latest/hadoop-provided.html

File: Dockerfile (hunk 2 above)
Comment (chrissng) [reply Dockerfile:87]: `export SPARK_DIST_CLASSPATH=$(hadoop classpath)` (the content in spark-env.sh) is already in `entrypoint.sh`

File: Dockerfile (hunk 1 above)
Comment (chrissng) [reply Dockerfile]: Removed and updated README with some instructions.

File: Dockerfile (hunk 2 above)
Comment (chrissng) [reply Dockerfile]: Fetching spark binaries without hadoop now.

Comment (chrissng) [reply Dockerfile:87]: Since the spark binary without hadoop pre-built does not come with hive, im just going to fetch this now.

**Comment by xkjyeah** [comment]
Hang on it does. Just tested 😶

**Comment by xkjyeah** [comment]
You might still need to add minimal configuration like so:

https://github.com/datagovsg/spark-yarn-cluster/tree/master/spark-conf

1. Set classpath
2. Set some `spark-defaults.conf`:
  - `spark.yarn.jars=file:///opt/spark-2.X/jars/*.jar`
  - `spark.master = yarn`

On the user's end, he/she needs to
1) Mount `hive-site.xml` onto spark/conf
2) Mount his hadoop configuration and set the ENV

**Comment by xkjyeah** [comment]
Ok my bad:

1. `without-hadoop` does *not* come with Hive.
2. `hadoop2.6` or `hadoop2.7` comes with Hive.

Sorry, let's revert to the `hadoop2.6` version -- it takes far less effort than having to build Spark manually.

**Review by xkjyeah — APPROVED** [review]

**Pull request merged**